File: src/main/java/io/protobase/f7/formulas/math/ACOT.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ACOT extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ACOT;
  public static ACOT SELF = new ACOT();

  public ACOT() {
    super();
  }

  public ACOT(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    double divisor = Converters.toDouble(collateralLookup.apply(origin, Converters.first(values[0])));
    if (divisor == 0 || divisor == -0) {
      return 1.570796327;
    }
    return Math.atan(1 / divisor);
  }
}

File: src/main/js/formulas/math/ACOT.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ACOT extends AbstractFormula {
  static SELF: ACOT = new ACOT();
  NAME = FormulaName.ACOT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const divisor = Converters.toNumber(this.collateralLookup(origin, Converters.first(values[0])));
    if (divisor === 0) {
      return 1.570796327;
    }
    return Math.atan(1 / divisor);
  }
}
